src/cfnlint/rules/resources/NoEcho.py

```diff
@@ -0,0 +1,43 @@
+"""
+Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+SPDX-License-Identifier: MIT-0
+"""
+from ast import literal_eval
+from cfnlint.rules import CloudFormationLintRule
+from cfnlint.rules import RuleMatch
+
+
+class NoEcho(CloudFormationLintRule):
+    id = 'W4002'
+    shortdesc = 'Check for NoEcho References'
+    description = 'Check if there is a NoEcho enabled parameter referenced within a resource'
+    source_url = 'https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#parameters-section-structure-properties'
+    tags = ['resources', 'NoEcho']
+
+    def match(self, cfn):
+        matches = []
+        no_echo_params = []
+        parameters = cfn.get_parameters()
+
+        for parameter_name, parameter_value in parameters.items():
+            parameter_properties = literal_eval(str(parameter_value))
+            for parameter_properties_name, parameter_properties_value in parameter_properties.items():
+                if parameter_properties_name == 'NoEcho' and parameter_properties_value is True:
+                    no_echo_params.append(parameter_name)
+
+        resource_properties = cfn.get_resources()
+
+        for resource_name, resource_values in resource_properties.items():
+
+            resource = literal_eval(str(resource_values))
+            for key, value in resource.items():
+                if key == 'Metadata':
+                    metadata = literal_eval(str(value))
+                    for prop_name, prop_value in metadata.items():
+                        path = ['Resources', resource_name, 'Metadata', prop_name]
+                        matches.append(RuleMatch(path, 'As the resource "metadata" section contains reference to a '
+                                                       '"NoEcho" parameter (' + prop_value + '), CFN will display the '
```

Comment:
[`exception while processing rule W4002: can only concatenate str (not "dict") to str`](https://travis-ci.com/github/aws-cloudformation/cfn-python-lint/jobs/355611684#L311)

https://github.com/aws-cloudformation/cfn-python-lint/blob/f209ddfef9bcc1a005adfebcfcc16220b18deddb/test/fixtures/templates/bad/core/directives.yaml#L7-L11

https://github.com/aws-cloudformation/cfn-python-lint/blob/f209ddfef9bcc1a005adfebcfcc16220b18deddb/test/fixtures/templates/bad/core/mandatory_checks.yaml#L7-L11

Reply:
Fixed in newer patches